pkg/network/server.go

```diff
@@ -387,10 +387,28 @@ func (s *Server) ConnectedPeers() []string {
 // while itself dealing with peers management (handling connects/disconnects).
 func (s *Server) run() {
 	go s.runProto()
-	for {
-		if s.PeerCount() < s.MinPeers {
+	for loopCnt := 0; ; loopCnt++ {
+		var (
+			netSize = s.discovery.NetworkSize()
+			// "Optimal" number of peers.
+			optimalN = s.discovery.GetFanOut() * 2
+			// Real number of peers.
+			peerN = s.PeerCount()
+		)
+
+		if peerN < s.MinPeers {
+			// Starting up or going below the minimum -> quickly get many new peers.
 			s.discovery.RequestRemote(s.AttemptConnPeers)
+		} else if s.MinPeers > 0 && loopCnt%s.MinPeers == 0 && optimalN > peerN && optimalN < s.MaxPeers && optimalN < netSize {
+			// Having some number of peers, but probably can get some more, the network is big.
+			// It also allows to start picking up new peers proactively, before we suddenly have <s.MinPeers of them.
+			var connN = s.AttemptConnPeers
+			if connN > optimalN-peerN {
+				connN = optimalN - peerN
+			}
+			s.discovery.RequestRemote(connN)
```

Comment:
Have you check this strategy on the network with a large number of available peers (maybe on mainnet o testnet)? What happens in the long run?

Reply:
It works.

Reply:
BTW, I've also looked at https://ethernodes.org/history --- E-letter chain has ~5-10K of nodes, so the optimal fan out for it (as we've defined it) is around 21-23 and it has these parameters:
```
./geth --maxpeers VALUE: set maximum number of full node peers, default 25
./geth --lightpeers VALUE: set maximum number of light node peers, default 100
```